examples/nodejs/storage/user-program.ts

```diff
@@ -0,0 +1,98 @@
+      // We can make it easy for the developer to exhaustively check the type
+      // and handle the cases they care about as follows:
+      switch (getResponse.type) {
+        case StoreValueType.STRING:
+          console.log('The value is a string:', getResponse.tryGetValueString());
+          break;
+        case StoreValueType.INTEGER:
+          console.log('The value is an integer:', getResponse.tryGetValueInteger());
+          break;
+        case StoreValueType.FLOAT:
+          console.log('The value is a float:', getResponse.tryGetValueFloat());
+          break;
+        case StoreValueType.BOOLEAN:
+          console.log('The value is a boolean:', getResponse.tryGetValueBoolean());
+          break;
+        default:
+          console.error('Unknown type');
+      }
+      // An alternative to this switch is:
+      // - StoreGet.Success has a `value` field which is of type `StoreValue`
+      // `StoreValue` is a base class with derived types, each with a `value` field.
+      // - StoreStringValue with a string-valued `value` field,
+      // - StoreIntegerValue with an integer-valued `value` field,
+      // - StoreFloatValue, etc.
+      // In that world the developer can use more instanceof checks:
+      // if (getResponse instanceof StoreGet.Success) {
+      //    const storeValue = getResponse.value;
+      //    if (storeValue instanceof StoreStringValue) {
+      //      console.log('The value is a string:', storeValue.value);
+      //    } else if (storeValue instanceof StoreIntegerValue) {
+      //      console.log('The value is an integer:', storeValue.value);
+      //    } else if (storeValue instanceof StoreFloatValue) {
+      //      // ...
+      //    } else {
+      //      console.error('Unknown type');
+      //    }
+      // }
+      //
+      // I feel this pattern makes sense in a language with first-class pattern matching, such as
+      // Rust, Scala, Kotlin, etc.
+      // But in JavaScript, I feel the switch is at the right level of complexity.
+      // This introduces an extra level of indirection where the dev has to access the `Success`
+      // value and then the `value` field of the `StoreValue` object.
+    } else if (getResponse instanceof StoreGet.Error) {
+      console.error('Error:', getResponse.message);
+    } else {
+      console.error('Unknown response type');
+    }
+  }
+
+  // I know what I'm doing
+  const response = await client.get(storeName, 'string-key');
+  if (response instanceof StoreGet.Success) {
+    console.log(`The value is: ${response.tryGetValueString()} which is the same as ${response.value as string}`);
+    console.log("If I don't know what I'm doing, I can catch the exception");
+    try {
+      console.log(response.tryGetValueInteger());
+    } catch (error: unknown) {
+      console.error(error);
+    }
+  } else if (response instanceof StoreGet.Error) {
+    console.error('Error:', response.message);
+  }
+}
+
+main().then().catch(console.error);
```

Comment:
thanks for putting this together. my other question is, as you were doing this, did you have any sense of whether this is going to feel good/bad/weird/different/better/worse vs. if we did not model the types directly? e.g. if the proto was just `bytes`? honestly on that front I feel like this exercise was a success, because we would still need to support `string | uint8Array` everywhere, and from my first read through this PR it seems like the challenges that we faced in surfacing that through the current CacheClient API are about the same as the challenges here (from an end-user type-checking perspective), and here we have stronger correctness guarantees. WDYT?

Reply:
> If nothing else you could move the .tryGetValue* methods up to the parent class so that they could be accessed directly on the response without having to do the instance check first?

I thought about this and didn't feel good about the user having to distinguish accessing the wrong type vs the panoply of other errors that could have happened. This may be the lowest friction UX though:

```typescript
const getResponse = await client.get('my-store', 'my-key');
try {
    return getResponse.tryGetValueString();
} catch (GetResponseException e) {
   switch (e.errorCode) {
        case GetResponse.ErrorCode.WrongType:
            // ...
        case GetResponse.ErrorCode.KEY_NOT_FOUND:
          // ...
        case GetResponse.ErrorCode.STORE_NOT_FOUND:
         // ...
   }
}
```

Reply:
> thanks for putting this together. my other question is, as you were doing this, did you have any sense of whether this is going to feel good/bad/weird/different/better/worse vs. if we did not model the types directly? e.g. if the proto was just bytes? honestly on that front I feel like this exercise was a success, because we would still need to support string | uint8Array everywhere, and from my first read through this PR it seems like the challenges that we faced in surfacing that through the current CacheClient API are about the same as the challenges here (from an end-user type-checking perspective), and here we have stronger correctness guarantees. WDYT?

I agree. Even if on the wire we only had bytes, users time and again have shown string is the default (possibly legacy from redis but that's another story). So we already have the problem of supporting more than one type on the SDK dev-facing API.

I really like the idea of having more types supported on the wire for portability. A developer could write an integer from a mobile frontend and read it from a web backend. If we support the types out of the box then they can write an integer directly without worrying about all the integer encoding nuances or marshalling to a neutral type and parsing themselves.

Reply:
> I'd say the store API has much less surface area than the cache

The Cache API started with only get/set. :)

The Store API is going to grow over time and very well may end up supporting the same types of collections and such. So we should make sure that the current minimal size of the API is not a precondition to us preferring the explicitly modeled data types :)